### DISCUSSION TITLE
Graal native images with supplied executables

### DISCUSSION THREAD

**Discussion by ennerf** [hydraulic-software/conveyor#19] [Q&A] [answered]
Out of curiousity, have you thought about integration with Graal native images yet? I'm not expecting cross-platform builds, but it would be cool to bundle some pre-compiled CLI tools.

For example, let's say I have multiple executables in one code base and I'd like to share a single native image for all of them. As far as I know Graal  is limited to a single class entry point and can't produce a DLL, so it'd have to generate some sort of launcher / forwarder where the first argument specifies the action.

It would be cool if I could supply a `native_image_launcher.exe` as input, and then have a conveyor-generated `cli_tool.exe` that calls it with something like `${APPDIR}/native_image_launcher.exe ${cli_tool_action} ${user-args}` . 

Would something like this already work? (I haven't looked at the C++ docs yet)

update:
It looks like this can already be done pretty easily with a generated `.exe` and a few shell/batch scripts.

**Comment by mrbubble** [comment, marked as answer]
Hi, Florian,

Yes, in principle native images should work as regular native apps without all the JVM specific stuff, but we haven't tested it yet. You might need some linker flags on macOS to ensure sufficient header space for code injection (mentioned in the [tutorial](https://conveyor.hydraulic.dev/1.0/tutorial/8-explore-the-integration/), see "Code injection on macOS").

By the way, you can make shared libraries with native images. At some point in the future we could make the launcher load a native image and map different EXEs to different main classes as we do now for regular JVM apps. That'd be useful for suites of CLI tools (for GUI apps you need a single entry point because of macOS), but it's currently a lower priority.

**Reply by ennerf** [reply]
Interesting. I'll have to look at that again 👍 Thanks.
